server/src/com/cloud/vm/UserVmManagerImpl.java

```diff
@@ -1228,6 +1230,19 @@ public UserVm addNicToVirtualMachine(AddNicToVMCmd cmd) throws InvalidParameterV
         return _vmDao.findById(vmInstance.getId());
     }
 
+    /**
+     * If the given MAC address is invalid it replaces the given MAC with the next available MAC address
+     */
+    protected String validateOrReplaceMacAddress(String macAddress, long networkId) {
+        if (!NetUtils.isValidMac(macAddress)) {
```

Comment:
I am having second thoughts, when a user entered a MAC address but made a typo, we are now creating a new address for them, uncalled for. shouldn't we have refused and thrown some horrible thing instead?

Reply:
I think we should only generate a MAC address if the value entered by the user is null. All other combinations (invalid, empty, blank and so on) should return an invalid parameter value exception.

Reply:
isn't empty or blank the same as null from a user perspective? It is up to the API-client how it passes no value. (we agree on invalid completely)

Reply:
Well.. yes and no. It depends on how the API is handling this case. I am not sure if all (`blank`, `empty`, `null`) will be represented as the same. I mean, I do not know if whenever we enter a blank or empty case, if these values will be treated as `null`  by ACS.